### PR TITLE
Push and pull to v2 registries with sha256 instead of tarsum

### PR DESCRIPTION
**Pull request by dmcgowan**
Adds digest package from distribution and uses it to generate sha256 digests on push and verify sha256 and tarsum on pull.

### CODE REVIEW

**Comment by tiborvass** [comment]
@dmcgowan rebase please :)

**Comment by icecrime** [comment]
@tiborvass This has a dependent PR#11143...

**Comment by dmcgowan** [comment]
Dependency merged and rebased

File: registry/digest/verifiers.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	case "sha256", "sha384", "sha512":
+		return hashVerifier{
+			hash:   newHash(alg),
+			digest: d,
+		}
+	default:
+		// Assume we have a tarsum.
+		version, err := tarsum.GetVersionFromTarsum(string(d))
+		if err != nil {
+			panic(err) // Always assume valid tarsum at this point.
```
Comment (jlhawn) [inline registry/digest/verifiers.go]: It looks like I can get the daemon to panic if I give it a digest that is not sha or tarsum?

**Comment by LK4D4** [comment]
@dmcgowan If digest package is part of distribution, why we can't vendor it, to have fixed version and easy updating?

File: registry/digest/verifiers.go (hunk 1 above)
Comment (dmcgowan) [inline registry/digest/verifiers.go]: Lets get the updates from distribution merged and I will update this PR

**Comment by jlhawn** [comment]
@dmcgowan I think it's okay to call `layer.SaveCheckSum()` on pull now.

**Comment by dmcgowan** [comment]
@LK4D4 agreed vendoring is better but I was trying to avoid pulling everything in.  Does this look better?
https://github.com/dmcgowan/docker/commit/c29f6d756d19a323d75ee36cff1423c937f01392

I will update the PR after we merge some updates upstream, it will address all the comments on it as well.

File: graph/push.go [hunk 2]
```diff
@@ -442,29 +443,22 @@ func (s *TagStore) pushV2Image(r *registry.Session, img *image.Image, endpoint *
 		os.Remove(tf.Name())
 	}()
 
-	ts, err := tarsum.NewTarSum(arch, true, tarsum.Version1)
+	h := sha256.New()
+	size, err := bufferToFile(tf, io.TeeReader(arch, h))
```
Comment (jlhawn) [inline graph/push.go]: @dmcgowan can we gzip this while we're at it?

Comment (dmcgowan) [inline graph/push.go]: Can we? yes, Should we? probably, but can we make a separate PR for it.  It will require additional testing and verifying with the registry that there are no problems.  We have no been using gzip on the engine side for the v2 registry thus far, but agree now is probably the best time to start.

Comment (jlhawn) [inline graph/push.go]: right - not necessary for this PR.

Once this is merged, I'll submit 2 separate PRs for caching checksums on pull and gzip of content on push.

**Comment by icecrime** [comment]
I'm +1 on the vendor-based solution!

**Comment by icecrime** [comment]
Considering the amount of tarsum related crap, I think everybody will agree on the need to switch to something else: I'm tentatively pushing from `1-design-review` to `2-code-review`, please correct if you feel differently!

**Comment by jlhawn** [comment]
@icecrime that seems reasonable, especially as most (if not all) of the discussion on this PR so far has been about the `digest` package and not about the changes to the `graph` package.

**Comment by jlhawn** [comment]
`graph` package changes LGTM, by the way. :+1:

**Comment by icecrime** [comment]
For my part, I reviewed the `graph` changes and everything looks good, I only gave the `digest` package an quick look.

**Comment by jlhawn** [comment]
@icecrime I've taken @tiborvass's comments and opened this PR on the distribution project from which it is vendored https://github.com/docker/distribution/pull/251

**Comment by dmcgowan** [comment]
Vendored digest from distribution after the change from @jlhawn was merged.

**Comment by jlhawn** [comment]
L<sup>G<sup>T<sup>M</sup></sup></sup>

**Comment by dmp42** [comment]
LGTM

**Comment by tiborvass** [comment]
I couldn't test it against a V1 registry for some technical reasons (`docker pull registry` never terminates for some reason :( )

Other than that, LGTM

**Comment by tiborvass** [comment]
@dmcgowan needs rebase

**Comment by dmcgowan** [comment]
Rebased

**Comment by tiborvass** [comment]
@dmcgowan sorry need another rebase :(

**Comment by dmcgowan** [comment]
Rebased again, unfortunately because of placement of the code in vendor.sh it comes across a merge conflict everyt ime libcontainer is updated

**Comment by dmcgowan** [comment]
Rebased, resolved changes with progress reader in master

**Comment by dmcgowan** [comment]
Rebased, resolved changes related to pull by digest

**Comment by jlhawn** [comment]
LGTM

**Comment by tiborvass** [comment]
I retested this with V1 registry, and v2 registry, and it all works. The additional UI change is the output `Digest: sha256:0206c748e849b67e95bb98521724819fbc67fa3c90df355b1b87ccb6314dd10e` on the command line. Re LGTM

**Comment by jfrazelle** [comment]
LGTM
